File: 0310/solution.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Integer> level = new ArrayList<>();
        if (n == 1) {
            level.add(0);
            return level;
        }
        int[] indegrees = new int[n];
        Map<Integer, List<Integer>> graph = new HashMap<>();
        for (int[] edge : edges) {
            int x = edge[0];
            int y = edge[1];
            indegrees[x]++;
            indegrees[y]++;
            graph.putIfAbsent(x, new ArrayList<>());
            graph.putIfAbsent(y, new ArrayList<>());
            graph.get(x).add(y);
            graph.get(y).add(x);
        }

        for (int i = 0; i < indegrees.length; i++) {
            if (indegrees[i] == 1) {
                level.add(i);
            }
        }

        while (true) {
            List<Integer> nextLevel = new ArrayList<>();
            for (int node : level) {
                for (int child : graph.get(node)) {
                    indegrees[child]--;
                    if (indegrees[child] == 1) {
                        nextLevel.add(child);
                    }
                }
            }
            if (nextLevel.isEmpty()) {
                return level;
            }
            level = nextLevel;
        }
    }
}

// time O(n)
// space O(n)

File: 0310/solution.ts
function findMinHeightTrees(n: number, edges: number[][]): number[] {
  let level: Array<number> = [];
  if (n == 1) {
    return [0];
  }
  const indegrees = new Array(n).fill(0);
  const graph = new Map<number, Array<number>>();
  for (const [x, y] of edges) {
    indegrees[x]++;
    indegrees[y]++;
    graph.set(x, graph.get(x) || []);
    graph.set(y, graph.get(y) || []);
    graph.get(x)!.push(y);
    graph.get(y)!.push(x);
  }

  for (let i = 0; i < indegrees.length; i++) {
    if (indegrees[i] == 1) {
      level.push(i);
    }
  }

  while (true) {
    const nextLevel: Array<number> = [];
    for (const node of level) {
      for (const child of graph.get(node)!) {
        indegrees[child]--;
        if (indegrees[child] == 1) {
          nextLevel.push(child);
        }
      }
    }
    if (!nextLevel.length) {
      return level;
    }
    level = nextLevel;
  }
}
